src/graph/hungarian-algorithm.md

```diff
@@ -0,0 +1,291 @@
+If we talk about the asymptotic complexity of the algorithm, then it is $\mathcal{O}(n^4)$: in total there should be at most $n$ increases in matching, before each of which there are no more than $n$ potential recalculations, each of which is performed in time $\mathcal{O}(n^2)$.
+
+We will not give the implementation for the $\mathcal{O}(n^4)$ algorithm here, since it will turn out to be no shorter than the implementation for the $\mathcal{O}(n^3)$ one, described below.
+
+### The $\mathcal{O}(n^3)$ algorithm
+
+Now let's learn how to implement the same algorithm in $\mathcal{O}(n^3)$ (for rectangular problems $n \times m$, $\mathcal{O}(n^2m)$).
+
+The key idea is to **consider matrix rows one by one**, and not all at once. Thus, the algorithm described above will take the following form:
+
+1.  Consider the next row of the matrix $A$.
+
+2.  While there is no increasing path starting in this row, recalculate the potential.
+
+3.  As soon as an augmenting path is found, alternate the matching along it (thus including the last row in the matching), and restart from point 1 (to consider the next line).
+
+To achieve the required asymptotic complexity, it is necessary to implement steps 2-3, which are performed for each row of the matrix, in time $\mathcal{O}(n^2)$ (for rectangular problems in $\mathcal{O}(nm)$).
+
+To do this, recall two facts proved above:
+
+- With a change in the potential, the vertices that were reachable by Kuhn's traversal will remain reachable.
+
+- In total, only $\mathcal{O}(n)$ recalculations of the potential could occur before an augmenting path was found.
+
+From this follow these **key ideas** that allow us to achieve the required asymptotics:
+
+- To check for the presence of an augmenting path, there is no need to start the Kuhn traversal again after each potential recalculation. Instead, you can make the Kuhn traversal in an **iterative form**: after each recalculation of the potential, look at the added rigid edges and, if their left ends were reachable, mark their right ends reachable as well and continue the traversal from them.
+
+- Developing this idea further, we can present the algorithm as follows: a cyclic process is employed where, at each step, the potential is recalculated. Subsequently, a column that has become reachable is identified (which will always exist as new reachable vertices emerge after every potential recalculation). If the column is unsaturated, an augmenting chain is discovered. Conversely, if the column is saturated, the corresponding row in the matching also becomes reachable.
+
+- To quickly recalculate the potential (faster than the $\mathcal{O}(n^2)$ naive version), you need to maintain auxiliary minima for each of the columns:
+
+$$minv[j]=\min_{i\in Z_1} A[i][j]-u[i]-v[j].$$
+
+It's easy to see that the desired value $\Delta$ is expressed in terms of them as follows:
```

Comment:
This, and a paragraph below should have the same tabulation as "- To quickly recalculate...", but it breaks because of block maths. It'd be better to fix this

Reply:
I added css to math blocks to fix tabulation. It's not so elegant but it works.